### PR TITLE
Refactor IOUModal to MoneyRequestModal

### PR DESCRIPTION
**Pull request by thienlnam**
### Fixed Issues

$ https://github.com/Expensify/Expensify/issues/270592
$ https://github.com/Expensify/App/issues/16276

### Tests

This is a refactor of the IOUModal so lets test the main flows in all sorts of ways

- Request Money
1. Request money via the global create
2. Update the currency to something random
3. Select a participant
4. Verify a transaction is created between you and the participant

- Split bill
1. Split bill via the global create
2. Update the currency to USD
3. Select 8 participants and notice you see the text `You've selected the maximum number (8) of participants.`
4. Continue and write a description
5. Notice that a group chat was created, and in each DM a message was sent

- Send Money
1. Send money via the global create
2. Update the currency to USD
3. Select a participant and hit 'Settle outside of Expensify'
4. Verify it works as expected

- Verify that the nothing about the IOU functionality has changed

- [ ] Verify that no errors appear in the JS console

### Offline tests

1. Go through all of the flows offline and then go back online and verify they work as expected

### QA Steps

Same steps as Tests

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/30609178/229651071-975a8129-26fa-44b2-aee6-973feb92e1a0.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/30609178/229885341-53850e6d-54db-43c4-8503-e52137176634.mov

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/30609178/229887266-a416562c-383f-4c3f-878b-19d4ab578055.mov

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/30609178/229659877-9e3e1c9a-6df1-4704-b62d-cdfef02a1ece.mov

</details>

<details>
<summary>iOS</summary>

</details>

<details>
<summary>Android</summary>

</details>

### CODE REVIEW

File: src/pages/iou/MoneyRequestModal.js
Comment (marcaaron) [inline src/pages/iou/MoneyRequestModal.js:160]: NAB / maybe for a follow up. But I would consider refactoring this logic into a custom hook called `useNetworkReconnect()` that does something like this:

exports the context object from here: https://github.com/Expensify/App/blob/4efde840cc7456d14a2f79185e95cd9d95f79b90/src/components/OnyxProvider.js#L9-L10

```js
function useNetworkReconnect(callback) {
    const {isOffline} = useContext(NetworkContext);
    const prevIsOfflineRef = useRef(isOffline);
    useEffect(() => {
        if (prevIsOfflineRef.current && !isOffline) {
            callback();
        }

        prevIsOfflineRef.current = isOffline;
    }, [isOffline]);
}
```

and would be used like this

```js
useNetworkReconnect(PersonalDetails.openMoneyRequestModalPage);
```

File: src/pages/iou/MoneyRequestModal.js
Comment (marcaaron) [inline src/pages/iou/MoneyRequestModal.js:168]: NAB, but feel like I don't understand it or maybe it's an anti-pattern. Why is the currency changing? Can we just update this whenever the currency changes (e.g. by moving `IOU.setIOUSelectedCurrency()` closer to the logic that updates the prop instead of applying a side effect)?

File: src/pages/iou/MoneyRequestModal.js
Comment (marcaaron) [inline src/pages/iou/MoneyRequestModal.js:170]: might be good to add a comment explaining this early return.

File: src/pages/iou/MoneyRequestModal.js
Comment (marcaaron) [inline src/pages/iou/MoneyRequestModal.js:180]: Might be better as two separate `useEffect()` (ignore if we take my advice above and create a custom hook for the network stuff)

File: src/pages/iou/MoneyRequestModal.js
Comment (marcaaron) [inline src/pages/iou/MoneyRequestModal.js:392]: Pretty much all of these functions should use `useCallback()` if they have dependencies.

**Review by marcaaron — COMMENTED** [review]
Looking pretty 👌. Left a few notes on the hooks stuff.

File: src/pages/iou/MoneyRequestModal.js
Comment (marcaaron) [reply src/pages/iou/MoneyRequestModal.js:392]: And if they don't we can move them to the top of the file. I looked briefly and I think most do have dependencies.

File: src/pages/iou/MoneyRequestModal.js
Comment (marcaaron) [inline src/pages/iou/MoneyRequestModal.js:351]: Thoughts on just creating a new "Header" component we can use for this? This feels a little "heavy" as an inline function and only has a few dependencies that can be passed as props.

File: src/pages/iou/MoneyRequestModal.js
Comment (thienlnam) [reply src/pages/iou/MoneyRequestModal.js:160]: I agree it would be nice to have a seperate hook for this, I'll create another issue that will go in and just refactor all instances of it

Comment (thienlnam) [reply src/pages/iou/MoneyRequestModal.js:160]: https://github.com/Expensify/App/issues/16889

File: src/pages/iou/MoneyRequestModal.js
Comment (thienlnam) [reply src/pages/iou/MoneyRequestModal.js:168]: Yeah I also agree this is an anti-pattern, I would imagine we call this where the logic happens of the user changing their currency code

File: src/pages/iou/MoneyRequestModal.js
Comment (thienlnam) [reply src/pages/iou/MoneyRequestModal.js:351]: In one of my commits, I started breaking this out into a separate header - I ended up moving it back since I had to pass a bunch of props that were very specific to this file and felt like it could be better generalized in a separate PR that refactors it

Comment (thienlnam) [reply src/pages/iou/MoneyRequestModal.js:351]: Added this issue for cleanup later https://github.com/Expensify/App/issues/16890

File: src/pages/iou/MoneyRequestModal.js
Comment (thienlnam) [reply src/pages/iou/MoneyRequestModal.js:168]: ... I went to the spot where the currency is selected and it already does this so I am going to 🔪 it from here and see what breaks

File: src/pages/iou/MoneyRequestModal.js
Comment (thienlnam) [reply src/pages/iou/MoneyRequestModal.js:338]: Actually, it's not that bad so will just do it here

File: src/pages/iou/MoneyRequestModal.js
Comment (thienlnam) [reply src/pages/iou/MoneyRequestModal.js]: What's the guideline for when to cache methods? I didn't think it was necessary here because none of these functions perform any heavy computational tasks that need to be cached

File: src/ROUTES.js
Comment (Julesssss) [inline src/ROUTES.js:11]: 👍

File: src/pages/iou/IOUSendPage.js
Comment (Julesssss) [inline src/pages/iou/IOUSendPage.js:6]: Is the newline for just 2 chars necessary here?

File: src/pages/iou/MoneyRequestModal.js
Comment (Julesssss) [inline src/pages/iou/MoneyRequestModal.js:32]: NAB: Also can be used for Sending money

File: src/pages/iou/MoneyRequestModal.js
Comment (Julesssss) [inline src/pages/iou/MoneyRequestModal.js:138]: Just out of curiousity, what is the linter complaining about here?

`react-hooks/exhaustive-deps`

File: src/pages/iou/ModalHeader.js
Comment (mountiny) [inline src/pages/iou/ModalHeader.js:13]:
```suggestion
    /** Title of the header */
```

**Review by mountiny — COMMENTED** [review]
Amazing job @thienlnam!

File: src/pages/iou/IOUSendPage.js
Comment (Julesssss) [reply src/pages/iou/IOUSendPage.js:6]: Oh I can't either anymore, ignore this.

File: src/pages/iou/MoneyRequestModal.js
Comment (thienlnam) [reply src/pages/iou/MoneyRequestModal.js:138]: We just want to run this on mount, and because I'm using `props.currentUserPersonalDetails.localCurrencyCode` within the effect it wants me to add it as a dependency - but it's not needed because we get it from onyx

**Comment by thienlnam** [comment]
Addressed review comments

File: src/pages/iou/MoneyRequestModal.js
Comment (marcaaron) [inline src/pages/iou/MoneyRequestModal.js:35]: nab, should we change some of these `IOU` to `Request`?

File: src/pages/iou/MoneyRequestModal.js
Comment (marcaaron) [reply src/pages/iou/MoneyRequestModal.js:138]: >it's not needed because we get it from onyx

And because we do not want this effect to run again.

I think it's kind of an anti-pattern though (that we can clean up later probably). If we always want the "iou selected currency" to be the "local currency code" when this component mounts then we should update that value in Onyx whenever we update the local currency code. We can hook into the initial render like this - but it's not obvious if we should...

File: src/pages/iou/MoneyRequestModal.js
Comment (marcaaron) [inline src/pages/iou/MoneyRequestModal.js:143]: Is it kind of sus and/or confusing? If the error occurs from the API then could we just look at `props.iou.error` as a dependency? If we didn't have an error before, but have one now then we should set the index to 0.

also having trouble understanding the case where we'd dismiss the modal. Anytime the `iou` changes and we have no errors and we're not in the process of creating the transaction via the API we call `dismissModal()`?

File: src/pages/iou/MoneyRequestModal.js
Comment (marcaaron) [inline src/pages/iou/MoneyRequestModal.js:175]: `useCallback()`

File: src/pages/iou/MoneyRequestModal.js
Comment (marcaaron) [inline src/pages/iou/MoneyRequestModal.js:194]: `useCallback()`

File: src/pages/iou/MoneyRequestModal.js
Comment (marcaaron) [inline src/pages/iou/MoneyRequestModal.js:227]: `useCallback()`

File: src/pages/iou/MoneyRequestModal.js
Comment (marcaaron) [inline src/pages/iou/MoneyRequestModal.js:239]: `useCallback()`

File: src/pages/iou/MoneyRequestModal.js
Comment (marcaaron) [inline src/pages/iou/MoneyRequestModal.js:253]: `useCallback()`

File: src/pages/iou/MoneyRequestModal.js
Comment (marcaaron) [inline src/pages/iou/MoneyRequestModal.js:298]: `useCallback()`

**Review by marcaaron — CHANGES_REQUESTED** [review]
Each time we render this component the inline functions with dependencies inside it will get re-created. Based on this conversation we have decided to just wrap those with `useCallback()`. 

Full context here: https://expensify.slack.com/archives/C01GTK53T8Q/p1680096510744619

File: src/pages/iou/MoneyRequestModal.js
Comment (thienlnam) [reply src/pages/iou/MoneyRequestModal.js:145]: Yeah agreed this doesn't make much sense in the hooks world. Going to make some changes to this

Comment (thienlnam) [reply src/pages/iou/MoneyRequestModal.js:145]: Yeah I agree it is kind of confusing

> If the error occurs from the API then could we just look at props.iou.error as a dependency?

We could, but we'd still need to check the previous props to see if we are done creating iou otherwise this would run each re-render

> also having trouble understanding the case where we'd dismiss the modal. Anytime the iou changes and we have no errors and we're not in the process of creating the transaction via the API we call dismissModal()? Why is that exactly?

Don't have a great answer for this, I'm mostly just refactoring the existing logic in place

Comment (thienlnam) [reply src/pages/iou/MoneyRequestModal.js:145]: My guess actually is that we have this logic in place for the offline handling of creating IOUs. Like if you dismiss the modal right after the transaction was created, then we don't have the ability to re-open and restart the money request. This handling is basically like - user comes back online? okay open the money request modal and then now that they are back online based on the success state of the IOU either restart the modal or dismiss it

File: src/pages/iou/MoneyRequestModal.js
Comment (thienlnam) [reply src/pages/iou/MoneyRequestModal.js]: Updated some of them to Request - didn't update all of them because they are still IOU reports and used like that throughout the code so didn't want it to cause more confusion

File: src/pages/iou/MoneyRequestModal.js
Comment (marcaaron) [reply src/pages/iou/MoneyRequestModal.js:145]: Ok, I think at a minimum we should just add a more specific dependency for `creatingIOUTransaction`? If I am understanding the logic correctly we only want to do "something" (show error or dismiss) when the transaction API call completes.

Comment (thienlnam) [reply src/pages/iou/MoneyRequestModal.js:145]: I believe we're already kind of handling this with prevCreatingIOUTransactionStatusRef, like we need to check that creatingIOUTransaction is true but also that the previous creatingIOUTransaction is false to confirm this is the first time that the IOU was created

Comment (thienlnam) [reply src/pages/iou/MoneyRequestModal.js:145]: More weirdness happening, creating this issue https://github.com/Expensify/App/issues/16942 to investigate it more deeply but not block on this

**Comment by thienlnam** [comment]
Okay, there's some more weird logic to clean up but I don't want to keep holding this PR for these changes bc many other PRs are on hold for this.

I created another issue to investigate more deeply here https://github.com/Expensify/App/issues/16942 but this will at least unblock the other PRs.  Right now, this uses mostly the same logic that was already existing

**Comment by thienlnam** [comment]
Ready to review / test!

File: src/pages/iou/MoneyRequestModal.js
Comment (marcaaron) [inline src/pages/iou/MoneyRequestModal.js:133]: Another possible follow up, but we should definitely use a lazy init for `useState()` here.. we need to pull the logic from line 109 down into the `useState()` call so we only calculate it once instead of on every render (might not be hurting anything but it's unnecessary work)...

```suggestion
    // Skip IOUParticipants step if participants are passed in
    const steps = reportParticipants.length ? [Steps.IOUAmount, Steps.IOUConfirm] : [Steps.IOUAmount, Steps.IOUParticipants, Steps.IOUConfirm];

    const prevCreatingIOUTransactionStatusRef = useRef(lodashGet(props.iou, 'creatingIOUTransaction'));
    const prevNetworkStatusRef = useRef(props.network.isOffline);

    const [previousStepIndex, setPreviousStepIndex] = useState(0);
    const [currentStepIndex, setCurrentStepIndex] = useState(0);
    const [participants, setParticipants] = useState(() => _.map(OptionsListUtils.getPersonalDetailsForLogins(reportParticipants, props.personalDetails), personalDetails => ({
        login: personalDetails.login,
        text: personalDetails.displayName,
        firstName: lodashGet(personalDetails, 'firstName', ''),
        lastName: lodashGet(personalDetails, 'lastName', ''),
        alternateText: Str.isSMSLogin(personalDetails.login) ? Str.removeSMSDomain(personalDetails.login) : personalDetails.login,
        icons: [{
            source: ReportUtils.getAvatar(personalDetails.avatar, personalDetails.login),
            name: personalDetails.login,
            type: CONST.ICON_TYPE_AVATAR,
        }],
        keyForList: personalDetails.login,
        payPalMeAddress: lodashGet(personalDetails, 'payPalMeAddress', ''),
        phoneNumber: lodashGet(personalDetails, 'phoneNumber', ''),
    })));
```

**Review by marcaaron — APPROVED** [review]
LGTM - full disclosure I did not test the changes yet, but will give it a go if needed.

**Review by fedirjh — APPROVED** [review]

File: src/pages/iou/MoneyRequestModal.js
Comment (thienlnam) [reply src/pages/iou/MoneyRequestModal.js:133]: Good point, will add that to the list of things to do in https://github.com/Expensify/App/issues/16942

**Comment by marcaaron** [comment]
Gonna merge this so we can keep the Manual Requests train rolling 🚅

File: src/pages/iou/ModalHeader.js
Comment (sobitneupane) [inline src/pages/iou/ModalHeader.js:31]: There was [some inconsistency](https://github.com/Expensify/App/issues/17366) in styling Header with Back button. It was fixed by [this PR](https://github.com/Expensify/App/pull/17560/files).

**Pull request merged**
